### PR TITLE
check upvars in closures that are in statics

### LINKED ISSUE #23057: ICE when save a closure to static mut produced from closure call
code sample:
```rust
static mut add2: Box<FnMut() -> i32> = Box::new((||{let mut a = 0; move || {a+=1;a}})());
```

compiler complans:
```
error: internal compiler error: unexpected panic
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'assertion failed: fcx.inh.deferred_call_resolutions.borrow().is_empty()', /home/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/librustc_typeck/check/vtable.rs:268

```

[playpen](http://is.gd/mb7JuU)

### LINKED ISSUE #27890: ICE when trying to make a static closure
After some discussing about static closures on IRC, I was messing around and found an ICE:

    static plus_one: &'static Fn(i32) -> i32 = (&|x: i32| { x + 1 }) as &'static Fn(i32) -> i32;
    

```
 error: internal compiler error: coherence failed to report ambiguity: cannot locate the impl of the trait `core::ops::Fn<(i32,)>` for the type `[closure@hello.rs:1:46: 1:64]`
hello.rs:1 static plus_one: &'static Fn(i32) -> i32 = (&|x: i32| { x + 1 }) as &'static Fn(i32) -> i32;
                                                      ^~~~~~~~~~~~~~~~~~~~~

thread 'rustc' panicked at 'Box<Any>', ../src/libsyntax/diagnostic.rs:176

stack backtrace:
   1:     0x7fb188388719 - sys::backtrace::write::h0942de78b6c02817K8r
   2:     0x7fb1883905a6 - panicking::on_panic::h3f23f9d0b5f4c91bu9w
   3:     0x7fb18835668e - rt::unwind::begin_unwind_inner::h2844b8c5e81e79558Bw
   4:     0x7fb185364717 - rt::unwind::begin_unwind::h18145712027808195836
   5:     0x7fb1853646d6 - diagnostic::SpanHandler::span_bug::h8e44e4841a51abaeu0A
   6:     0x7fb18634d25a - middle::traits::error_reporting::report_fulfillment_errors::h333f72cab9295933RyR
   7:     0x7fb1873e70c5 - check::FnCtxt<'a, 'tcx>::select_all_obligations_or_error::hcab104bc898397548ur
   8:     0x7fb1873e4b83 - check::regionck::Rcx<'a, 'tcx>::visit_region_obligations::hfa6295200248132aJBc
   9:     0x7fb1873e65f0 - check::regionck::Rcx<'a, 'tcx>::visit_fn_body::hef046c27196f1a04oyc
  10:     0x7fb1873ec73e - visit::walk_expr::h8219604655069971555
  11:     0x7fb1873e9ca6 - check::regionck::visit_expr::h7ba995ff7bd03dafSXc
  12:     0x7fb1873e9ee5 - check::regionck::visit_expr::h7ba995ff7bd03dafSXc
  13:     0x7fb1873ea018 - check::regionck::visit_expr::h7ba995ff7bd03dafSXc
  14:     0x7fb1873ec31c - visit::walk_expr::h8219604655069971555
  15:     0x7fb1873e9a8d - check::regionck::visit_expr::h7ba995ff7bd03dafSXc
  16:     0x7fb1873e4778 - check::regionck::regionck_expr::h31bdba3ccd579413Aoc
  17:     0x7fb18748b19b - check::check_const::hea0d071c5597530dvcu
  18:     0x7fb18747941e - check::check_item_type::haa6e26c4456d14ddoKp
  19:     0x7fb18753a1e2 - check_crate::hc9d8771979c9f6f1ZzE
  20:     0x7fb1888d2ffe - driver::phase_3_run_analysis_passes::closure.20596
  21:     0x7fb1888b1e05 - middle::ty::ctxt<'tcx>::create_and_enter::h9049188843559879450
  22:     0x7fb1888acf5d - driver::phase_3_run_analysis_passes::h10313786347949334605
  23:     0x7fb188890676 - driver::compile_input::h68da15dd21986404Tba
  24:     0x7fb1889fb7eb - run_compiler::h707077bbe9100331Obc
  25:     0x7fb1889f90b7 - boxed::F.FnBox<A>::call_box::h6790869519760498004
  26:     0x7fb1889f8b24 - rt::unwind::try::try_fn::h1844525071116418565
  27:     0x7fb188390148 - __rust_try
  28:     0x7fb18837ca62 - rt::unwind::try::inner_try::h45702ba1ad1620941xw
  29:     0x7fb1889f8cb8 - boxed::F.FnBox<A>::call_box::h11351074772415591031
  30:     0x7fb18838f593 - sys::thread::Thread::new::thread_start::h66ead7b2912d4f30OHv
  31:     0x7fb181fa30a3 - start_thread
  32:     0x7fb18801204c - clone
  33:                0x0 - <unknown>

```

### LINKED ISSUE #28099: ICE at static boxed mutable function
I tried this code:

```rust
static no_op : Box<FnMut(i32)> = Box::new(move |_| { });
fn main() {
}
```

I expected it to compile cleanly or give me a helpful error.  Instead, it gave me a helpful error and then an ICE:

```
src/bug_repro.rs:1:34: 1:56 error: function calls in statics are limited to struct and enum constructors [E0015]
src/bug_repro.rs:1 static no_op : Box<FnMut(i32)> = Box::new(move |_| { });
                                                    ^~~~~~~~~~~~~~~~~~~~~~
src/bug_repro.rs:1:34: 1:56 help: run `rustc --explain E0015` to see a detailed explanation
error: internal compiler error: unexpected panic
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'assertion failed: self.mode == Mode::Var', /home/rustbuild/src/rust-buildbot/slave/stable-dist-rustc-linux/build/src/librustc/middle/check_const.rs:254
```

Version:

```
$ rustc --version --verbose
rustc 1.1.0 (35ceea399 2015-06-19)
binary: rustc
commit-hash: 35ceea3997c79a3b7562e89b462ab76af5b86b22
commit-date: 2015-06-19
host: x86_64-unknown-linux-gnu
release: 1.1.0
```

stack backtrace:

```
   1:     0x7f87cd852529 - sys::backtrace::write::ha3b8868b13c6f5b594r
   2:     0x7f87cd85a449 - panicking::on_panic::h0d4f0a7bb69824a9iJw
   3:     0x7f87cd81b242 - rt::unwind::begin_unwind_inner::h6eae1ff9ccde81e1sow
   4:     0x7f87cb69a14e - rt::unwind::begin_unwind::h5657407561811364434
   5:     0x7f87cb7e0352 - middle::check_const::CheckCrateVisitor<'a, 'tcx>.Visitor<'v>::visit_fn::hd06dbbc963009a38mQd
   6:     0x7f87cb7e327b - visit::walk_expr::h7428632329866803358
   7:     0x7f87cb7daf58 - middle::check_const::CheckCrateVisitor<'a, 'tcx>.Visitor<'v>::visit_expr::ha2cfc0edff92e762pSd
   8:     0x7f87cb7dae1e - middle::check_const::CheckCrateVisitor<'a, 'tcx>.Visitor<'v>::visit_expr::ha2cfc0edff92e762pSd
   9:     0x7f87cb7c6150 - middle::check_const::CheckCrateVisitor<'a, 'tcx>::global_expr::h84b8593855198411OFd
  10:     0x7f87cb7df439 - middle::check_const::CheckCrateVisitor<'a, 'tcx>.Visitor<'v>::visit_item::h07c7afb9c4056bb8tLd
  11:     0x7f87cb7e799a - middle::check_const::check_crate::h91fd962e256ef08cnke
  12:     0x7f87cdda222e - driver::phase_3_run_analysis_passes::h9b9e8dc3e9bbf431tGa
  13:     0x7f87cdd8606c - driver::compile_input::h6e596544d26b4394Qba
  14:     0x7f87cde3c041 - run_compiler::haa6c0289fab7415475b
  15:     0x7f87cde39892 - boxed::F.FnBox<A>::call_box::h10721825727138786690
  16:     0x7f87cde38e59 - rt::unwind::try::try_fn::h6014602198652783625
  17:     0x7f87cd8d2b58 - rust_try_inner
  18:     0x7f87cd8d2b45 - rust_try
  19:     0x7f87cde390f4 - boxed::F.FnBox<A>::call_box::h5510246712457526188
  20:     0x7f87cd8591e1 - sys::thread::Thread::new::thread_start::h11fbc90f6ba541f9zvv
  21:     0x7f87c80dd6a9 - start_thread
  22:     0x7f87cd4a2eec - clone
  23:                0x0 - <unknown>

```

### LINKED ISSUE #28113: ICE in rustc from function call in RHS of const definition
I tried this code:

    const X: u8 = || -> u8 { 5 }();

I expected to see this happen: Either a compilation error, or to be equivalent to const X: u8 = 5.

Instead, this happened: Internal compiler error

## Meta

`rustc --version --verbose`:

```
rustc 1.4.0-nightly (20a8412e0 2015-08-28)
binary: rustc
commit-hash: 20a8412e09e2883489736b4ae3dc138f243d1a17
commit-date: 2015-08-28
host: x86_64-unknown-linux-gnu
release: 1.4.0-nightly
```

Backtrace: 

```
thread 'rustc' panicked at 'assertion failed: self.inh.deferred_call_resolutions.borrow().is_empty()', ../src/librustc_typeck/check/mod.rs:2023

stack backtrace:
   1:     0x7fc7fbaf90a9 - sys::backtrace::tracing::imp::write::he11991336d112ae3Aes
   2:     0x7fc7fbb00d76 - panicking::on_panic::hd741984fb33f3e317ex
   3:     0x7fc7fbac55de - rt::unwind::begin_unwind_inner::h837b3c3463e1a06eLHw
   4:     0x7fc7fab0ca9f - rt::unwind::begin_unwind::h10622579821064218427
   5:     0x7fc7fab3216c - check::FnCtxt<'a, 'tcx>::select_all_obligations_or_error::h966bbb42357b6836dvr
   6:     0x7fc7fabd43fc - check::check_const::hbed0eb360d678b26Acu
   7:     0x7fc7fabc22c7 - check::check_item_type::h25c10dc6ab99c072tKp
   8:     0x7fc7fac84392 - check_crate::hf540b00aced9403awAE
   9:     0x7fc7fc049a8e - driver::phase_3_run_analysis_passes::closure.20656
  10:     0x7fc7fc02c0c3 - middle::ty::ctxt<'tcx>::create_and_enter::h8301109304510330010
  11:     0x7fc7fc02745d - driver::phase_3_run_analysis_passes::h13747097652036984327
  12:     0x7fc7fc00a876 - driver::compile_input::h0626ad86fff9a5b9Tba
  13:     0x7fc7fc16c98b - run_compiler::h4c5bab99983d75bd0bc
  14:     0x7fc7fc16a257 - boxed::F.FnBox<A>::call_box::h7998076066074844987
  15:     0x7fc7fc169cc4 - rt::unwind::try::try_fn::h11048267709636862519
  16:     0x7fc7fbb00918 - __rust_try
  17:     0x7fc7fbaed1d2 - rt::unwind::try::inner_try::h62b49309b595e978EDw
  18:     0x7fc7fc169e58 - boxed::F.FnBox<A>::call_box::h8109257951937123469
  19:     0x7fc7fbaffd63 - sys::thread::Thread::new::thread_start::h78cb9ca46e8b6bb6rNv
  20:     0x7fc7f56c0181 - start_thread
  21:     0x7fc7fb77947c - __clone
  22:                0x0 - <unknown>

```

### PR DESCRIPTION
**Pull request by arielb1**
Fixes #23057 
Fixes #27890
Fixes #28099
Fixes #28113

r? @pnkfelix

### CODE REVIEW

File: src/test/compile-fail/borrowck-in-static.rs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// check that borrowck looks inside consts/statics
```
Comment (arielb1) [inline src/test/compile-fail/borrowck-in-static.rs]: Now that this code type-checks, it would be a shame if borrowck doesn't actually look inside.

**Comment by alexcrichton** [comment]
ping r? @pnkfelix

**Comment by alexcrichton** [comment]
triage: beta-nominated

@arielb1 indicates that the regression reported in https://github.com/rust-lang/rust/issues/28801 was fixed by this PR (e.g. working in 1.3, broken in 1.4 (beta), and working in 1.5), so nominating for a backport to beta.

**Comment by pnkfelix** [comment]
@rust-lang/compiler I'm in favor of backporting this PR to beta.

**Comment by nikomatsakis** [comment]
Accepted as fixes a regression.

**Pull request merged**